text/0000-stable-simd.md

```diff
@@ -0,0 +1,720 @@
+allowing `stdsimd` to live on crates.io) we'll instead pull in `stdsimd` to the
+standard library and expose it as the stable interface to SIMD in Rust.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+There's a number of unresolved questions around stabilizing SIMD today which
+don't pose serious blockers and may also wish to be considered open bugs rather
+than blocking stabilization:
+
+## Relying on unexported LLVM APIs
+
+The static detection performed by `cfg!` and `#[cfg]` currently relies on a
+[Rust-specific patch to LLVM][llvm-patch]. LLVM internal knows all about
+hierarchies of features and such. For example if you compile with `-C
+target-feature=+avx2` then `cfg!(target_feature = "sse2")` also needs to resolve
+to `true`. Rustc, however, does not know about these features and relies on
+learning this information through LLVM.
+
+Unfortunately though LLVM does not actually export this information for us to
+consume (as far as we know). As a result we have a [local patch][llvm-patch]
+which exposed this information for us to read. The consequence of this
+implementation detail is that when compiled against the system LLVM the `cfg!`
+macro may not work correctly when used in conjunction with `-C target-feature`
+or `-C target-cpu` flags.
+
+It appears that clang [vendors and/or duplicates][clang] LLVM's functionality in
+this regard. It's an option for rustc to do the same but it may also be an
+option to expose the information in upstream LLVM. So far there appears to have
+been no attempts to upstream this patch into LLVM itself.
+
+[llvm-patch]: https://github.com/rust-lang/llvm/commit/68e1e29618b2bd094d82faac16cf8e89959bbd68
+[clang]: https://github.com/llvm-mirror/clang/blob/679d846fcc73bd213347785185006d591698a132/lib/Basic/Targets/X86.cpp
+
+## Intrinsics that take constant arguments and function pointers
+
+Mentioned above it's planned to give intrinsics that take constant arguments
+some special treatment, but unfortunately we still haven't forbidden function
+pointers to these intrinsics. That means that it's still possible to pass
+runtime-determined values which may otherwise not be possible to do in a world
+where it's powered by `const` features rather than special attributes.
+
+## Packed SIMD types in `extern` functions are not sound
+
+The packed SIMD types have particular care paid to them with respect to their
+ABI in Rust and how they're passed between functions, notably to ensure that
+they work properly throughout Rust programs. The "fix" to pass them in memory
+over function calls, however, was only applied to the "Rust" ABI and not any
+other function ABIs.
+
+A consequence of this change is that if you instead label all your functions
+`extern` then the [same bug][soundbug] will arise. It may be possible to
+implement a "lint" or a compiler error of sorts to forbid this situation in the
+short term. We could also possibly accept this as a known bug for the time
+being.
```

Comment:
We have this lint already, don't we? The `simd_ffi` feature gate. I'd be in favor of leaving that as-is.

(Though it should be extended to [interact correctly with repr(transparent)](https://play.rust-lang.org/?gist=4a6bd3a25ebe5379fe00fa04e092a4b9&version=nightly). Not critical right now since SIMD types are unstable too, but if both are stabilized...)

Reply:
Aha indeed! I think it may need to be extended to general `extern` functions though rather than just those in `extern { ... }` blocks. (I haven't tested it much myself though)

Reply:
I started to take a stab at implementing this and in the process realized a problem: While foreign functions can't be generic and SIMD types can therefore be detected reliably, Rust-defined functions can be generic with any ABI. So I guess we can't actually enforce this except with a monomorphization time error? 🙄

Reply:
Heh yeah that sounds like a problem! I'd personally be ok with a sort of "best effort lint" as we really just want to give a heads up that it won't work, and typically you're not using generics in extern fns

Reply:
Couldn't we make it a compile time error to call an `extern "C"` function with a SIMD parameter from a function that isn't explicitly tagged with the required `#[target_feature]` (if the base target doesn't already include the feature).

Reply:
@parched Not pre-monomorphization. Since generic `extern` functions are valid today, so is this code:

```
extern fn bar<T>(x: T) {}
fn foo<T>(x: T) { bar(x); }
```

Reply:
Ah yes, sorry I didn't quite read your comment correctly. I guess one (not perfect) solution would be to not actually follow the C ABI for generic functions and pass SIMD types on the stack like `extern "Rust"` as, in theory, that would be unobservable.

Reply:
It would be observable in that it would break function pointers: `let f: fn(__m128) = bar::<__m128>;` (assuming `extern fn bar<T>(x: T) { ... }`).

Treating generic functions differently from monomorphic ones would also strongly go against how the compiler operates today.